{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calcAndAppendStatValForScan(alleleCounts, snpLocs, statName, subWinStart, subWinEnd, hapsInSubWin, statVals):\n",
    "# modified code from https://github.com/kern-lab/diploSHIC/blob/master/fvTools.py\n",
    "    if statName == \"tajD\":\n",
    "        statVals[statName].append(allel.stats.diversity.tajima_d(\n",
    "            alleleCounts, pos=snpLocs, start=subWinStart, stop=subWinEnd))\n",
    "    elif statName == \"pi\":\n",
    "        statVals[statName].append(allel.stats.diversity.sequence_diversity(\n",
    "            snpLocs, alleleCounts, start=subWinStart, stop=subWinEnd))\n",
    "    elif statName == \"thetaW\":\n",
    "        statVals[statName].append(allel.stats.diversity.watterson_theta(\n",
    "            snpLocs, alleleCounts, start=subWinStart, stop=subWinEnd)) \n",
    "    elif statName == 'H2/H1' or statName == 'H12':\n",
    "        h1,h12,h123,h21 = allel.stats.selection.garud_h(hapsInSubWin)\n",
    "        if statName == 'H2/H1':\n",
    "            statVals[statName].append(h21)\n",
    "        else:\n",
    "            statVals[statName].append(h12)\n",
    "    else:\n",
    "        print(statName + \" not found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel\n",
    "from allel.model.ndarray import SortedIndex\n",
    "from allel.util import asarray_ndim\n",
    "import re\n",
    "import subprocess\n",
    "import os.path\n",
    "\n",
    "datadir = \"/media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/\"\n",
    "cmd      = 'ls ' + datadir + '*.vcf.gz'\n",
    "fileList =  subprocess.run(cmd, shell = True, stdout=subprocess.PIPE).stdout.decode('utf-8')\n",
    "fileList = fileList.split(\"\\n\")\n",
    "fileList = list(filter(None, fileList))\n",
    "\n",
    "winsize = 1000     # how big should the sliding window be around each SNP?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10900_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10901_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10902_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10903_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10904_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10905_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10906_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10907_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10908_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10909_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10910_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10911_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10912_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10913_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10914_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10915_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10916_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10917_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10918_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10919_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10920_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10922_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10923_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10924_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10925_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10927_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10928_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10929_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10930_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10931_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10932_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10933_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10934_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10935_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10936_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10937_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10938_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10939_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10940_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10941_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10942_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10943_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10944_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10945_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10946_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10947_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10948_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10949_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10950_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10951_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10952_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10953_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10954_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10955_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10956_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10957_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10958_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10959_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10960_Invers_VCFallFILT.vcf.gz\n",
      "Processing /media/kevin/TOSHIBA_EXT/TTT_RecombinationGenomeScans/results_final/10961_Invers_VCFallFILT.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "for f in fileList:\n",
    "    print(\"Processing \" + f)\n",
    "    vcf     = allel.read_vcf(f, fields = [\"CHROM\", \"POS\", \"GT\"])\n",
    "    m       = re.search('[0-9]{5}', f)      # m is a regex match object\n",
    "    simNum  = m.group(0)                    # sim number is the first set of 5 numbers in the vcf name\n",
    "    \n",
    "    g       = allel.GenotypeArray(vcf[\"calldata/GT\"])\n",
    "    ac      = g.count_alleles()\n",
    "    haps    = g.to_haplotypes()\n",
    "    pos     = vcf['variants/POS']\n",
    "    \n",
    "    statvals = { \"tajD\"   : [],\n",
    "                 \"pi\"     : [],\n",
    "                 \"thetaW\" : [],\n",
    "                 \"H12\"    : [],\n",
    "                 \"H2/H1\"  : []}\n",
    "\n",
    "\n",
    "    for SNP in pos:\n",
    "        winStart = SNP - int(winsize/2)\n",
    "        winEnd   = SNP + int(winsize/2)\n",
    "        \n",
    "        if winStart < pos[0]:\n",
    "            winStart = pos[0]\n",
    "        \n",
    "        ####### subset the haplotype array ##############\n",
    "        startInd = np.searchsorted(pos, winStart) \n",
    "        endInd   = np.searchsorted(pos, winEnd, side = 'right') - 1 \n",
    "        # subtract 1 so the end index will not be outside the window. Side = 'right' in case the match is exact,\n",
    "        # in that case searchsorted will return the index after the match so the -1 will still be valid\n",
    "        \n",
    "        hapsInSubWin = allel.HaplotypeArray(haps.subset(list(range(startInd, endInd + 1))))\n",
    "    \n",
    "        for statName in statvals.keys():\n",
    "            calcAndAppendStatValForScan(ac, pos, statName, winStart, winEnd, hapsInSubWin, statvals)\n",
    "                                            \n",
    "    scanResultsFile = datadir + simNum + \"_Invers_ScanResults.txt\"\n",
    "    outfile         = datadir + simNum + \"_Invers_ScanResults_new.txt\"\n",
    "    \n",
    "    allFeatures = []\n",
    "    with open (scanResultsFile, 'r') as f:\n",
    "        for line in f:\n",
    "            features = line.split()\n",
    "            features = [f.strip('\"') for f in features]\n",
    "            allFeatures.append(features)\n",
    "        stats  = allFeatures[1:]\n",
    "        header = allFeatures[0]\n",
    "        featDf =  pd.DataFrame(stats, columns = header)\n",
    "        featDf = featDf[featDf.keep_loci == \"TRUE\"].reset_index(drop = True)\n",
    "\n",
    "    featDf = featDf.join(pd.DataFrame.from_dict(statvals))\n",
    "    featDf.to_csv(outfile, sep = \" \", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
